### ISSUE TITLE
Inconsistent handling of move constructor and move assignment in recursive_wrapper

### ISSUE THREAD

**Issue by joto** [mapbox/variant#50]
Move constructor nulls out operand but move assignment does a swap. See: https://github.com/mapbox/variant/blob/master/test/t/recursive_wrapper.cpp#L68-L85

I suggest using `swap` on both cases. This way the internal pointer can never be `nullptr` which saves us from an "undefined" state.

@artemp @springmeyer opinions?

**Comment by artemp** [comment]
@joto - not sure. R-value is 'undefined` after move op by definition so either is correct but swap will be less efficient.  (?)

**Comment by joto** [comment]
I just realized setting the `operand.p_` to `nullptr` leaks the memory allocated in it, so that's definitely not okay. If we use `swap` in both cases, we get defined behaviour and no leaks cheaply. If we use `nullptr`, we get a bunch of special cases which might be more efficient or not, hard to say. But definitely more complex code. In any case this should be consistent for move constructorand move assignment, otherwise that would be very suprising for the user.

**Comment by springmeyer** [comment]
@joto @artemp  - is this still relevant? What are the current questions and next actions?

**Comment by joto** [comment]
It think this is still relevant. I will look at it tomorrow.

**Comment by joto** [comment]
I changed the implementation of the move constructor to what is used in `boost::recursive_wrapper` in this case. I believe this is the right implementation, because it does a move construction of the underlying type, so it is as close as possible to that type.

**Comment by lightmare** [comment]
I'm confused, what was the problem? The 4 lines removed in b3df87f6de5ba5f5955558c37adce3c39e022b1b were perfectly fine, IMO, They were simply moving `operand.p_` to `this->p_`. The new line allocates a copy of an rvalue, seems unnecessary.

**Comment by lightmare** [comment]
The tests were incorrect in that they tested a moved-from value. But implementation was fine.

**Comment by joto** [comment]
I agree, the tests were incorrect, that's why I removed them. I am not saying the former implementation was wrong, but, as I explained above, I think the new one is better.

@artemp: do you have an opinion here?

**Comment by lightmare** [comment]
Compare:
```c++
unique_ptr<int> pi = unique_ptr<int>(new int(5));
recursive_wrapper<int> wi = recursive_wrapper<int>(5);
```
The former implementation did 1 allocation and moved the pointer from the temporary, just like unique_ptr. The new does a second allocation and copies the value.

**Comment by lightmare** [comment]
One last note, this could've been `noexcept(true)`:
```c++
    recursive_wrapper(recursive_wrapper && operand)
        : p_(operand.p_)
    {
        operand.p_ = nullptr;
    }
```
This can't:
```c++
    recursive_wrapper(recursive_wrapper && operand)
        : p_(new T(std::move(operand.get()))) {}
```
Anyway, #89 is superior in every way.

**Comment by tomilov** [comment]
Only (possible) downside is a [requirement to deleter](http://en.cppreference.com/w/cpp/memory/unique_ptr/~unique_ptr) (hereby destructor) should be noexcept.

**Comment by tomilov** [comment]
I think it is unlikely that someone will use throwing destructors. It is [very complicated matter](https://akrzemi1.wordpress.com/2011/09/21/destructors-that-throw/).

**Comment by lightmare** [comment]
I just realized something and need to refute what I wrote previously. The patch indeed fixes undefined behaviour.

> @lightmare The 4 lines removed in b3df87f were perfectly fine, IMO, They were simply moving operand.p_ to this->p_

I was wrong on this one. They were not fine, because they didn't leave the moved-from wrapper in a valid state -- destructible, yes, but any assignment to it, with the exception of move-assignment, would blow:
```c++
rwi x = 5;
rwi y = std::move(x); // x.p_ = nullptr
x = 6; // *(int*)nullptr = 6
```
Of course that comes down to the implementation of assigment operators, which would all need to check and allocate if `p_ == nullptr`. The committed solution is simpler.
